[DOCS] Release notes for v7.16.0 (#81369)
elastic/elasticsearch: docs/reference/release-notes/7.16.asciidoc

* Add utility for tracking licensed persistent tasks {es-pull}76672[#76672]
* Separate feature usage tracking for FLS and DLS {es-pull}79152[#79152]
* Use a licensed feature per realm-type (+custom) {es-pull}78810[#78810]

Machine Learning::
* Add new `normalize_above` parameter to `p_value` significant terms heuristic {es-pull}78833[#78833]
* Add new default char filter `first_line_with_letters` for machine learning categorization {es-pull}77457[#77457]
* Adding new `defer_definition_decompression` parameter to put trained model API {es-pull}77189[#77189] (issue: {es-issue}77132[#77132])
* Enable ML on macOS on ARM {es-pull}78203[#78203]
* Track feature usage for data frame analytics, inference, and anomaly jobs {es-pull}76789[#76789]

Mapping::
* Better error message for long keys in flattened fields {es-pull}80433[#80433] (issue: {es-issue}78248[#78248])
* Add `time_series_metric` parameter {es-pull}76766[#76766] (issue: {es-issue}74014[#74014])
* Add dimension mapping parameter {es-pull}74450[#74450] (issue: {es-issue}74014[#74014])

Monitoring::
* Remove license check for monitoring data retention {es-pull}79010[#79010]

Other::

> * Fleet: Add `action_response` into .fleet-actions-results mapping {es-pull}79584[#79584]

Reviewer: This looks to update an index template for a Fleet system data stream. I think it's a better fit under `Indices APIs`.
```suggestion
* Fleet: Add `action_response` into `.fleet-actions-results` mapping {es-pull}79584[#79584]
```

> * Filtering setting deprecation info API messages based on a setting {es-pull}78725[#78725]

Reviewer: "Other" seems like an odd label for this. I wonder if we want to relabel this as deprecations and reorder it on the page?

Packaging::
* Switch to Ubuntu docker base image {es-pull}80640[#80640]
* Use Cloudflare's zlib in Docker images {es-pull}81245[#81245] (issue: {es-issue}81208[#81208])
* Use almalinux as the Docker base image {es-pull}80524[#80524] (issue: {es-issue}76681[#76681])

Recovery::
* Add support for peer recoveries using snapshots after primary failovers {es-pull}77420[#77420] (issue: {es-issue}73496[#73496])
* Respect generational files in `recoveryDiff` {es-pull}77695[#77695] (issues: {es-issue}55142[#55142], {es-issue}55239[#55239])
* Limit concurrent snapshot file restores in recovery per node {es-pull}79316[#79316] (issue: {es-issue}79044[#79044])

Search::

> * Add `_ignored` and `_routing` metatada fields to fields api {es-pull}78981[#78981] (issues: {es-issue}75836[#75836], {es-issue}78828[#78828])
> * Add `_index` and `_version` metatada fields to fields api {es-pull}79042[#79042] (issues: {es-issue}75836[#75836], {es-issue}78828[#78828])

Reviewer: ```suggestion
* Add `_ignored` and `_routing` metatada fields to fields API {es-pull}78981[#78981] (issues: {es-issue}75836[#75836], {es-issue}78828[#78828])
* Add `_index` and `_version` metatada fields to fields API {es-pull}79042[#79042] (issues: {es-issue}75836[#75836], {es-issue}78828[#78828])
```

* Add ability to retrieve `_id` via fields option {es-pull}78828[#78828] (issue: {es-issue}75836[#75836])
* Add node-level field caps requests {es-pull}79212[#79212] (issues: {es-issue}74648[#74648], {es-issue}77047[#77047], {es-issue}78647[#78647])
* Add segment sorter for data streams {es-pull}75195[#75195]
* Add sort optimization with after from Lucene {es-pull}64292[#64292]
* Don't always rewrite the Lucene query in search phases {es-pull}79358[#79358]
* Expand segment sorter for all timeseries indices {es-pull}78639[#78639] (issue: {es-issue}75195[#75195])
* Node level can match action {es-pull}78765[#78765]

> * Search - return ignored field values from fields api {es-pull}78697[#78697] (issue: {es-issue}74121[#74121])

Reviewer: ```suggestion
* Search - return ignored field values from fields API {es-pull}78697[#78697] (issue: {es-issue}74121[#74121])
```

* Support request cache on frozen tier {es-pull}77694[#77694] (issue: {es-issue}75309[#75309])
* Use `search_coordination` threadpool in field caps {es-pull}79378[#79378] (issue: {es-issue}79212[#79212])
* Create a sha-256 hash of the shard request cache key {es-pull}74877[#74877] (issue: {es-issue}74061[#74061])

Security::
* Add `extensionName()` to security extension {es-pull}79329[#79329]
* Optimize `StringMatcher` for match-all patterns {es-pull}77738[#77738]

> * `CreateApiKey` response now returns the base64 encoded credentials {es-pull}77351[#77351] (issue: {es-issue}50235[#50235])

Reviewer: ```suggestion
* `CreateApiKey` response now returns the base64-encoded credentials {es-pull}77351[#77351] (issue: {es-issue}50235[#50235])
```

Snapshot/Restore::

> * Add Filtering by SLM Policy to Get Snapshots API {es-pull}77321[#77321]
> * Add Sort By Shard Count and Failed Shard Count to Get Snapshots API {es-pull}77011[#77011]

Reviewer: ```suggestion
* Add filtering by SLM policy to get snapshots API {es-pull}77321[#77321]
* Add sort by shard count and failed shard count to get snapshots API {es-pull}77011[#77011]
```

* Add descriptions to various tasks {es-pull}76700[#76700]
* Add maintenance service to clean up unused docs in snapshot blob cache {es-pull}77686[#77686]
* Add periodic maintenance task to clean up unused blob store cache docs {es-pull}78438[#78438] (issue: {es-issue}77686[#77686])
* Filter Unneeded `SnapshotInfo` Instances Early in `TransportGetSnapshotsAction` {es-pull}78032[#78032] (issue: {es-issue}74350[#74350])

> * Implement Exclude Patterns for Snapshot- and Repository Names in Get Snapshots API {es-pull}77308[#77308]
> * Implement Sort By Repository Name in Get Snapshots API {es-pull}77049[#77049]
> * Implement `from_sort_value` Parameter in Get Snapshots API {es-pull}77618[#77618]

Reviewer: ```suggestion
* Implement exclude patterns for snapshot and repository names in get snapshots API {es-pull}77308[#77308]
* Implement sort by repository name in get snapshots API {es-pull}77049[#77049]
* Implement `from_sort_value` parameter in get snapshots API {es-pull}77618[#77618]
```

Stats::
* Add cluster applier stats {es-pull}77552[#77552]
* Limit count of HTTP channels with tracked stats {es-pull}77303[#77303]
* Speedup computing cluster health {es-pull}78969[#78969] (issue: {es-issue}77466[#77466])
* Handle cgroups v2 in `OsProbe` {es-pull}77128[#77128] (issues: {es-issue}76812[#76812], {es-issue}77126[#77126])

Transform::
* Add `_meta` field to `TransformConfig` {es-pull}79003[#79003] (issue: {es-issue}77506[#77506])
* Add method to collect deprecations from a transform configuration {es-pull}77565[#77565]
* Add transform upgrade endpoint {es-pull}77566[#77566]
* Reduce indexes to query based on checkpoints {es-pull}75839[#75839]
* Implement the ability to preview the existing transform {es-pull}76697[#76697] (issue: {es-issue}76427[#76427])

[[bug-7.16.0]]
[float]
=== Bug fixes

Aggregations::
* Add extra round trip to aggs tests {es-pull}79638[#79638] (issue: {es-issue}73680[#73680])

> * Fix rate agg with custom `_doc_count` {es-pull}79346[#79346] (issue: {es-issue}77734[#77734])

Reviewer: ```suggestion
* Fix rate aggregation with custom `_doc_count` {es-pull}79346[#79346] (issue: {es-issue}77734[#77734])
```

> * Fix several potential circuit breaker leaks in Aggregators {es-pull}79676[#79676]

Reviewer: ```suggestion
* Fix several potential circuit breaker leaks in aggregators {es-pull}79676[#79676]
```

CRUD::
* Use query param instead of a system property for opting in for new cluster health response code {es-pull}79351[#79351] (issues: {es-issue}70849[#70849], {es-issue}78940[#78940])

Cluster Coordination::
* Avoid early release of local forking requests {es-pull}77641[#77641] (issues: {es-issue}77407[#77407], {es-issue}77634[#77634])
* Check for global blocks after `IndexNotFoundException` in `TransportMasterNodeAction` {es-pull}78128[#78128] (issue: {es-issue}70572[#70572])
* Improve control of outgoing connection lifecycles {es-pull}77295[#77295] (issue: {es-issue}67873[#67873])
* Only remove active peer on connection failure {es-pull}79557[#79557] (issues: {es-issue}77295[#77295], {es-issue}79550[#79550])
* Reduce merging in `PersistedClusterStateService` {es-pull}79793[#79793] (issue: {es-issue}77466[#77466])

CompositeAggs::
* Fix composite aggregation tests failing after #76740 {es-pull}77691[#77691] (issues: {es-issue}76740[#76740], {es-issue}77650[#77650])
* Revert 74559 (Avoid global ordinals in composite) {es-pull}78846[#78846] (issues: {es-issue}74559[#74559], {es-issue}78836[#78836])

Data streams::

> * Add replicated field to get data stream api response {es-pull}80988[#80988] (issue: {es-issue}118899[#118899])

Reviewer: ```suggestion
* Add replicated field to get data stream API response {es-pull}80988[#80988] (issue: {es-issue}118899[#118899])
```

> * Correct check for write index and increment generation on all DS backing index operations {es-pull}79916[#79916]

Reviewer: ```suggestion
* Correct check for write index and increment generation on all data stream backing index operations {es-pull}79916[#79916]
```

> * Fix Datastream Bug Causing it to Rollover to a non-existent Index {es-pull}79759[#79759]

Reviewer: ```suggestion
* Fix data stream bug causing it to rollover to a non-existent Index {es-pull}79759[#79759]
```

* Fix `IndexNotFoundException` error when handling remove alias action {es-pull}80312[#80312]
* Fix data stream alias validation {es-pull}81040[#81040] (issue: {es-issue}80972[#80972])

Distributed::

> * Fix fleet search API with no checkpoints {es-pull}79400[#79400]

Reviewer: ```suggestion
* Fix Fleet search API with no checkpoints {es-pull}79400[#79400]
```

> * Modify fleet search URLs to avoid url collisions {es-pull}79776[#79776]

Reviewer: ```suggestion
* Modify Fleet search URLs to avoid URL collisions {es-pull}79776[#79776]
```

EQL::
* Add optional fields and limit joining keys on non-null values only {es-pull}79677[#79677]

Geo::
* Fix bug filtering collinear points on vertical lines {es-pull}81155[#81155] (issues: {es-issue}59501[#59501], {es-issue}81076[#81076])
* Spherical mercator transformation should handle properly out of bounds latitudes {es-pull}81145[#81145] (issue: {es-issue}81128[#81128])

> * Vector tiles: Add key bucket value to the aggs layer {es-pull}79634[#79634] (issue: {es-issue}79585[#79585])

Reviewer: ```suggestion
* Vector tiles: Add key bucket value to the aggregation layer {es-pull}79634[#79634] (issue: {es-issue}79585[#79585])
```

ILM+SLM::

> * Prevent Duplicate ILM Cluster State Updates from Being Created {es-pull}78390[#78390] (issues: {es-issue}77466[#77466], {es-issue}78246[#78246])

Reviewer: ```suggestion
* Prevent duplicate ILM cluster state updates from being created {es-pull}78390[#78390] (issues: {es-issue}77466[#77466], {es-issue}78246[#78246])
```

> * Run ILM and SLM stopping cluster state updates at IMMEDIATE priority {es-pull}80207[#80207] (issue: {es-issue}80099[#80099])

Reviewer: ```suggestion
* Run ILM and SLM stopping cluster state updates at `IMMEDIATE` priority {es-pull}80207[#80207] (issue: {es-issue}80099[#80099])
```

* Validate that snapshot repository exists for ILM policies during `GenerateSnapshotNameStep` {es-pull}77657[#77657] (issue: {es-issue}72957[#72957])

Indices APIs::

> * Get-templates APIs don't support lists {es-pull}78989[#78989] (issue: {es-issue}78829[#78829])

Reviewer: ```suggestion
* Get templates APIs don't support lists {es-pull}78989[#78989] (issue: {es-issue}78829[#78829])
```

Infra/Core::
* Avoid spurious deprecation warnings when calling Deprecation Info API {es-pull}78151[#78151] (issue: {es-issue}78098[#78098])
* Fix overflow/underflow in `CompositeBytesReference` {es-pull}78893[#78893]
* Fix race condition in Feature Migration Status API {es-pull}80572[#80572] (issue: {es-issue}79680[#79680])
* Prevent stack overflow in rounding {es-pull}80450[#80450]

> * Set LIBFFI_TMPDIR at startup {es-pull}80651[#80651] (issues: {es-issue}18272[#18272], {es-issue}73309[#73309], {es-issue}74545[#74545], {es-issue}77014[#77014], {es-issue}77053[#77053], {es-issue}77285[#77285], {es-issue}80617[#80617])

Reviewer: ```suggestion
* Set `LIBFFI_TMPDIR` at startup {es-pull}80651[#80651] (issues: {es-issue}18272[#18272], {es-issue}73309[#73309], {es-issue}74545[#74545], {es-issue}77014[#77014], {es-issue}77053[#77053], {es-issue}77285[#77285], {es-issue}80617[#80617])
```

* Strip index blocks from settings for reindex targets {es-pull}80887[#80887] (issue: {es-issue}80654[#80654])
* Eschew leniency when parsing time zones {es-pull}77267[#77267] (issues: {es-issue}73955[#73955], {es-issue}76415[#76415])
* Fix ingest timezone parsing {es-pull}63876[#63876] (issue: {es-issue}63458[#63458])

Infra/Logging::
* Disable deprecation log indexing until templates are loaded {es-pull}80406[#80406] (issue: {es-issue}80265[#80265])

Infra/Settings::

> * [7.x] Stricter `UpdateSettingsRequest` parsing on the REST layer {es-pull}79228[#79228] (issue: {es-issue}29268[#29268])

Reviewer: ```suggestion
* Stricter `UpdateSettingsRequest` parsing on the REST layer {es-pull}79228[#79228] (issue: {es-issue}29268[#29268])
```

* Fix autoscaling capacity consistency {es-pull}81181[#81181]
* Fix bug in inference stats persister for when feature reset is called {es-pull}77213[#77213] (issue: {es-issue}77182[#77182])
* Fix datafeed preview with remote indices {es-pull}81099[#81099] (issue: {es-issue}77109[#77109])
* Fix language identification bug when multi-languages are present {es-pull}80675[#80675]
* Fix model snapshot sorting when sorting by `min_version` {es-pull}80596[#80596] (issue: {es-issue}80561[#80561])
* ML legacy index templates that are no longer needed should be deleted {es-pull}80874[#80874] (issue: {es-issue}80876[#80876])
* Need to tolerate .ml-config being an alias {es-pull}80025[#80025]
* Parent datafeed actions to the datafeed's persistent task {es-pull}81143[#81143]
* Wait for .ml-state-write alias to be readable {es-pull}79731[#79731] (issue: {es-issue}79636[#79636])

Mapping::
* Disable request cache for non-deterministic runtime fields {es-pull}75054[#75054]
* Fix `TextFieldMapper` Retaining a Reference to its Builder {es-pull}77251[#77251] (issue: {es-issue}73845[#73845])

Packaging::

> * Ensure LIBFFI_TMPDIR is exported by init script {es-pull}80794[#80794]

Reviewer: ```suggestion
* Ensure `LIBFFI_TMPDIR` is exported by init script {es-pull}80794[#80794]
```

Recovery::
* Do not release snapshot file download permit during recovery retries {es-pull}79409[#79409] (issue: {es-issue}79316[#79316])

SQL::

> * Fix NULLS FIRST/LAST for aggregations {es-pull}77750[#77750] (issue: {es-issue}34550[#34550])

Reviewer: ```suggestion
* Fix `NULLS FIRST/LAST` for aggregations {es-pull}77750[#77750] (issue: {es-issue}34550[#34550])
```

* Fix use of `requestTimeout` and `pageTimeout` query parameters {es-pull}79360[#79360] (issue: {es-issue}72151[#72151])
* Swap JDBC `page.timeout` and `query.timeout` properties in query requests {es-pull}79491[#79491] (issue: {es-issue}79430[#79430])

Search::
* Minimize search source of shard level search requests {es-pull}80634[#80634] (issue: {es-issue}80187[#80187])
* Prevent `NullPointerException` in `SourceConfirmedTextQuery` {es-pull}80472[#80472] (issue: {es-issue}80419[#80419])

> * Search bug fix - `_terms_enum` API `index_filter` doesn’t work with `_tier` field on upgraded cluster {es-pull}79553[#79553] (issue: {es-issue}79200[#79200])

Reviewer: ```suggestion
* `_terms_enum` API `index_filter` doesn’t work with `_tier` field on upgraded cluster {es-pull}79553[#79553] (issue: {es-issue}79200[#79200])
```

License::
* Add license family attribute to feature usage tracking {es-pull}76622[#76622]
* Add utility for tracking licensed persistent tasks {es-pull}76672[#76672]
* Separate feature usage tracking for FLS and DLS {es-pull}79152[#79152]
* Use a licensed feature per realm-type (+custom) {es-pull}78810[#78810]

Machine Learning::
* Add new `normalize_above` parameter to `p_value` significant terms heuristic {es-pull}78833[#78833]
* Add new default char filter `first_line_with_letters` for machine learning categorization {es-pull}77457[#77457]
* Add new `defer_definition_decompression` parameter to put trained model API {es-pull}77189[#77189] (issue: {es-issue}77132[#77132])
* Enable ML on macOS on ARM {es-pull}78203[#78203]
* Track feature usage for data frame analytics, inference, and anomaly jobs {es-pull}76789[#76789]
* Speed up training of regression and classification models {ml-pull}2024[#2024]
* Improve concurrency for training regression and classification models {ml-pull}2031[#2031]
* Improve aspects of implementation of `skip_model_update` rule {ml-pull}2053[#2053]

> * Make sure instrumentation captures the best hyperparameters found for training classification and regression models{ml-pull}2057{#2057}

Reviewer 2: ```suggestion
* Make sure instrumentation captures the best hyperparameters found for training classification and regression models {ml-pull}2057{#2057}
```

* Audit job open failures and stop any corresponding datafeeds {es-pull}80665[#80665] (issues: {es-issue}48934[#48934], {es-issue}80621[#80621])
* Fix acceptable model snapshot versions in ML deprecation checker {es-pull}81060[#81060] (issues: {es-issue}79387[#79387], {es-issue}81039[#81039], {es-issue}119745[#119745])
* Fix autoscaling capacity consistency {es-pull}81181[#81181]
* Fix bug in inference stats persister for when feature reset is called {es-pull}77213[#77213] (issue: {es-issue}77182[#77182])
* Fix datafeed preview with remote indices {es-pull}81099[#81099] (issue: {es-issue}77109[#77109])
* Fix language identification bug when multi-languages are present {es-pull}80675[#80675]
* Fix model snapshot sorting when sorting by `min_version` {es-pull}80596[#80596] (issue: {es-issue}80561[#80561])
* ML legacy index templates that are no longer needed should be deleted {es-pull}80874[#80874] (issue: {es-issue}80876[#80876])
* Need to tolerate .ml-config being an alias {es-pull}80025[#80025]
* Parent datafeed actions to the datafeed's persistent task {es-pull}81143[#81143]
* Wait for .ml-state-write alias to be readable {es-pull}79731[#79731] (issue: {es-issue}79636[#79636])

> * Correct ANOVA for Gaussian Process that is fit to the loss surface, which affects early stopping. Also, improve the estimates of hyperparameter importance {ml-pull}2073[#2073]

Reviewer 2: ```suggestion
* Improve the estimates of hyperparameter importance in data frame analytics and avoid incorrectly stopping the hyperparameter search prematurely {ml-pull}2073[#2073]
```

* Fix numerical instability in hyperparameter optimization for training regression and classification models {ml-pull}2078[#2078]

> * Fix numerical stability issues in time series modelling {ml-pull}2083[#[2083]]

Reviewer 2: ```suggestion
* Fix numerical stability issues in time series modeling {ml-pull}2083[#[2083]]
```

I think the US spelling has one L?  (Only accept suggestion if this is true.)
